src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs

```diff
@@ -624,22 +624,44 @@ private bool TryToRunInSTAThread(Action action, bool waitForCompletion)
         }
 
 
-        private static void UpdateTestResults(IEnumerable<TestResult> testResults, IEnumerable<TestCase> testCases, string package)
+        private static ICollection<TestCase> UpdateTestResultsAndInProgressTests(IEnumerable<TestResult> testResults, ICollection<TestCase> inProgressTestCases, string package)
         {
+
+            // No change required to testcases and testresults.
+            if (string.IsNullOrEmpty(package))
+            {
+                return inProgressTestCases;
+            }
+
+            EqtTrace.Verbose("BaseRunTests.UpdateTestResultsAndInProgressTests: Update source details for testResults and testCases.");
+
             // Before sending the testresults back, update the test case objects with source provided by IDE/User.
-            if (!string.IsNullOrEmpty(package))
+            foreach (var tr in testResults)
             {
-                foreach (var tr in testResults)
-                {
-                    tr.TestCase.Source = package;
-                }
+                tr.TestCase.Source = package;
```

Comment:
>tr.TestCase.Source = package; [](start = 16, length = 29)

Is this safe ?

Reply:
No, this not safe. As we discussed offline,  currently there is no known scenario to which break  because of not copying(better performance). However it is better to copy the test results(which adapter creates) to make sure correctness over performance.

I have made changes to copy the test results too.